app/src/test/java/com/example/sharingang/LiveDataTestUtil.kt

```diff
@@ -0,0 +1,41 @@
+package com.example.sharingang
+
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Observer
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+@VisibleForTesting(otherwise = VisibleForTesting.NONE)
+fun <T> LiveData<T>.getOrAwaitValue(
+    time: Long = 2,
+    timeUnit: TimeUnit = TimeUnit.SECONDS,
+    afterObserve: () -> Unit = {}
+): T {
+    var data: T? = null
+    val latch = CountDownLatch(1)
+    val observer = object : Observer<T> {
+        override fun onChanged(o: T?) {
+            data = o
+            latch.countDown()
+            this@getOrAwaitValue.removeObserver(this)
+        }
+    }
+    this.observeForever(observer)
+
+    try {
+        afterObserve.invoke()
+
+        // Don't wait indefinitely if the LiveData is not set.
+        if (!latch.await(time, timeUnit)) {
+            throw TimeoutException("LiveData value was never set.")
+        }
+
+    } finally {
+        this.removeObserver(observer)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    return data as T
+}
```

Comment:
I don't really understand how this class works or what its purpose is. What is it useful for? Could you please describe the big picture in 1-2 sentences?

Reply:
So, you use `LiveData` in your ViewModel instead of just a normal class, because you want the UI to be able to listen to updates / changes in the data.

The problem with testing this directly, is that now you have asynchronicity, because setting a new value isn't guaranteed to be immediately visible, because LiveData is designed to be used in the setting of an actual app.

This provides a `.getOrAwaitValue` method for the class `LiveData`, in order to block the current thread, waiting for the latest change. This exists only to make testing these classes possible, allowing you to turn `LiveData<T>` into `T` by blocking, and waiting for the latest result.

In all honesty, I don't understand how this snippet of code works, it was in the guide I read about Android testing:
https://developer.android.com/codelabs/advanced-android-kotlin-training-testing-basics#8

I think this is essentially the recommended way of testing `LiveData`, but it's not exactly pretty, and would probably be worthy of including into the standard Android extensions, but I digress.